Add inductor standalone_compile API

oulgen opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #150670

This PR adds standalone_compile API that does precompilation via caching to support vLLM use case in the short term while we work on the longer term precompilation solution.

```
standalone_compile(gm, example_inputs, options) -> CompiledArtifact
CompiledArtifact.save(path, format: binary|unpacked = binary)
CompiledArtifact.load(path, format: binary|unpacked = binary)
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

oulgen committed:
inductor_compile

[ghstack-poisoned]

oulgen committed:
Update on "inductor_compile"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

oulgen committed:
None on "Add inductor standalone_compile API"


This PR adds standalone_compile API that does precompilation via caching to support vLLM use case in the short term while we work on the longer term precompilation solution.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

aorenste commented on torch/_inductor/compile_fx.py:
```diff
@@ -2362,3 +2363,109 @@ def _aoti_flatten_inputs(
+) -> CompiledArtifact:
+    from torch.fx.experimental.symbolic_shapes import ShapeEnv
+
+    shape_env = ShapeEnv()
```
Consider not creating the default until you've determined that one wasn't provided in the example_inputs (a helper function would be useful to avoid Optional)

aorenste commented on torch/compiler/_cache.py:
```diff
@@ -121,6 +121,31 @@ def clear(cls) -> None:
 
+    @classmethod
+    def with_fresh_cache(cls) -> Any:
+        class FreshCache:
```
nit: consider using a `@contextlib.contextmanager`

oulgen replied:
what makes contextmanager better than this?

aorenste commented on torch/compiler/_cache.py:
```diff
@@ -121,6 +121,31 @@ def clear(cls) -> None:
         cls._cache_info.clear()
 
+    @classmethod
+    def with_fresh_cache(cls) -> Any:
```
Technically
```suggestion
    def with_fresh_cache(cls) -> AbstractContextManager[None]:
```

oulgen commented on torch/_inductor/compile_fx.py:
```diff
@@ -2362,3 +2363,109 @@ def _aoti_flatten_inputs(
+
+
+def standalone_compile(
+    gm: GraphModule, example_inputs: Sequence[InputType], **kwargs: Any
```
Do we want to expose compilefxkwargs type in init file though?

oulgen replied:
Actually I dont think this is true
compile_fx's signature is
```
def compile_fx(
    model_: GraphModule,
    example_inputs_: Sequence[InputType],
    inner_compile: Callable[..., OutputCode] = compile_fx_inner,
    config_patches: Optional[dict[str, Any]] = None,
    decompositions: Optional[dict[OpOverload, Callable[..., Any]]] = None,
) -> Union[Callable[[list[object]], Sequence[torch.Tensor]], str, list[str]]:
```

aorenste replied:
This is a good argument for using a real type instead of just `**kwargs: Any`. Otherwise someone looking at the code will have to read the function body to figure out how to call this / what valid args are...

oulgen committed:
None on "Add inductor standalone_compile API"


This PR adds standalone_compile API that does precompilation via caching to support vLLM use case in the short term while we work on the longer term precompilation solution.

```
standalone_compile(gm, example_inputs, options) -> CompiledArtifact
CompiledArtifact.save(filename)
CompiledArtifact.load(filename)
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

zou3519 commented:
cc @youkaichao if you had thoughts

zou3519 commented on torch/_inductor/compile_fx.py:
```diff
@@ -2362,3 +2367,109 @@ def _aoti_flatten_inputs(
+    from torch._subclasses import FakeTensorMode
+
+    shape_env = get_shape_env_from_example_inputs(example_inputs)
+    torch._guards._TLS.tracing_context = torch._guards.TracingContext(
+        FakeTensorMode(shape_env=shape_env)
+    )
```
there has to be a less hacky way of doing this lol. at least we need to unset the tracing_context when we are done with it

oulgen replied:
yea i can do this via a contextmanager

zou3519 commented on torch/_inductor/config.py:
```diff
@@ -117,6 +117,9 @@ def prologue_fusion_enabled() -> bool:
 force_disable_caches: bool = os.environ.get("TORCHINDUCTOR_FORCE_DISABLE_CACHES") == "1"
 
+# Unsafe way to skip dynamic shape guards to get faster cache load
+unsafe_skip_cache_dynamic_shape_guards: bool = False
```
Btw, how do the dynamic shape guards in FXGraphCache (?) work in general? Does one cache key refer to multiple artifacts and we need to use the dynamic shape guards to figure out which?

oulgen replied:
if there are inductor time guards that were added (which we wouldnt know at cache load time), then one key might refer to multiple artifacts, otherwise dynamo guards are part of cache key.

zou3519 commented on torch/_functorch/_aot_autograd/autograd_cache.py:
```diff
@@ -487,6 +487,9 @@ class AOTAutogradCacheEntry:
     forward_time_taken_ns: int
     backward_time_taken_ns: int
 
+    # Used by standalone_compile
+    sanitized_aot_config: AOTConfig
```
To check, these are now always included as a part of the cache key?

oulgen replied:
This is cache output, not key

zou3519 commented on torch/_inductor/codecache.py:
```diff
@@ -1021,16 +1021,20 @@ def iterate_over_candidates() -> Generator[
-                hit,
-            )
+            if config.unsafe_skip_cache_dynamic_shape_guards:
+                hit = True
```
maybe assert that there is only a single candidate?

oulgen replied:
I guess i was thinking we could use this flag for other cases in the future too where there might be more than one candidate

zou3519 reviewed (commented):
seems reasonable, have some questions

zou3519 commented on torch/_inductor/compile_fx.py:
```diff
@@ -2362,3 +2367,109 @@ def _aoti_flatten_inputs(
+                file.write(writer.to_bytes())
+
+    @staticmethod
+    def load(filename: str, *example_args: Any) -> CompiledArtifact:
```
Do we really need example_inputs to load from the cache?

oulgen replied:
wrap post compile uses it here: https://github.com/pytorch/pytorch/blob/0ad2c5d7e2e0ee6c847e167c2e2d65ec8d7321fa/torch/_functorch/_aot_autograd/autograd_cache.py#L558

not sure whether we can avoid it or not cc: @jamesjwu

jamesjwu replied:
It's using it because FXGraphCache requires example inputs for dynamic shape guards: if you made FXGraphCache not look at dynamic shape gaurds, you should be able to pass in an empty list to args

oulgen replied:
ok perfect, dead

jamesjwu replied:
Uh actually, you might need them for cudagraphs unfortunately; it's unclear to me how cudagraphs uses it, but it seems necessary there (see cudagraphs_post_compile in output_code.py)

zou3519 commented on torch/_inductor/compile_fx.py:
```diff
@@ -2362,3 +2367,109 @@ def _aoti_flatten_inputs(
+    from torch.compiler._cache import CacheArtifactManager
+
+    with CacheArtifactManager.with_fresh_cache():
+        compiled_fn = compile_fx(gm, example_inputs, **kwargs)
```
How realistic is it to error if Inductor tries to add any additional dynamic shape guards? cc @aorenste I remember you asked something similar a while back?

oulgen replied:
tensor size under 32 bits versus tensor size over 32 bits

jamesjwu replied:
There's a unittest in codecache.py that exhibits this behavior

oulgen commented on torch/_inductor/compile_fx.py:
```diff
@@ -2362,3 +2367,109 @@ def _aoti_flatten_inputs(
+        self._artifacts = artifacts
+
+    def __call__(self, *args: Any) -> Any:
+        return self._compiled_fn(*args)[0]
```
@zou3519 notice how i return [0] here, i assume for training there might be two functions here? what api do we want to provide?

zou3519 replied:
Why does `self._compiled_fn(*args)` return more than one output?

Also I thought for training there is only one AOTAutograd cache entry (but two Inductor FX graph cache entries) but the single callable from AOTAutograd cache wraps both

oulgen committed:
None on "Add inductor standalone_compile API"


This PR adds standalone_compile API that does precompilation via caching to support vLLM use case in the short term while we work on the longer term precompilation solution.

```
standalone_compile(gm, example_inputs, options) -> CompiledArtifact
CompiledArtifact.save(filename)
CompiledArtifact.load(filename, example_inputs)
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

oulgen committed:
None on "Add inductor standalone_compile API"


This PR adds standalone_compile API that does precompilation via caching to support vLLM use case in the short term while we work on the longer term precompilation solution.

```
standalone_compile(gm, example_inputs, options) -> CompiledArtifact
CompiledArtifact.save(filename)
CompiledArtifact.load(filename)
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

oulgen committed:
None on "Add inductor standalone_compile API"


This PR adds standalone_compile API that does precompilation via caching to support vLLM use case in the short term while we work on the longer term precompilation solution.

```
standalone_compile(gm, example_inputs, options) -> CompiledArtifact
CompiledArtifact.save(filename)
CompiledArtifact.load(filename)
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

jamesjwu reviewed (approved):
This LGTM from AOTAutogradCache side, besides the questions around example_inputs and cudagraphs. 

As a nit, it feels like the PR is missing a bunch of comments and docstrings for the main APIs (especially to go over the exact process of load() and save()), but you can add that before landing.

oulgen commented on torch/_inductor/compile_fx.py:
```diff
@@ -2362,3 +2368,136 @@ def _aoti_flatten_inputs(
+                files = list(os.listdir(autograd_cache_dir))
+                assert len(files) == 1
+                key = files[0]
+                cache_dir_ctx = temporary_cache_dir(path)
```
this sounds gross, but i guess i can change like a return value

zou3519 commented on torch/_inductor/compile_fx.py:
```diff
@@ -2362,3 +2368,136 @@ def _aoti_flatten_inputs(
+                with open(path, "wb") as file:
+                    file.write(writer.to_bytes())
+            else:
+                assert format == "unpacked"
+                assert os.path.isdir(path)
```
We probably should do a rm on the directory first? It's like how in the binary path the writer overwrites the file (does it?)

zou3519 commented on torch/_inductor/compile_fx.py:
```diff
@@ -2362,3 +2368,136 @@ def _aoti_flatten_inputs(
+
+                assert entry is not None
+
+                fx_config = _CompileFxKwargs(
+                    cudagraphs=BoxedBool(False),
+                    boxed_forward_device_index=BoxedDeviceIndex(0),
+                )
```
I guess we are punting on figuring out what these are?

zou3519 commented on torch/_inductor/compile_fx.py:
```diff
@@ -2362,3 +2368,136 @@ def _aoti_flatten_inputs(
     return flat_example_inputs, options
+
+
+_ENCODING_VERSION: int = 1
```
Btw, did we actually want a magic number instead of an encoding version? The summary from the group on Monday was that the artifact should only be usable if the pytorch version is the same, and the `_ENCODING_VERSION` sounds like a guardsafe if the user loads something that isn't the binary we produced

oulgen replied:
what's the difference between encoding version and magic number? we would need something to show that it is current source control version like torch_key() not version

zou3519 replied:
torch_key would show current source code version.

magic number is more of a "we trust the user that they are not abusing this API".

torch_key sounds better here

oulgen commented on test/inductor/test_codecache.py:
```diff
@@ -1326,6 +1327,63 @@ def forward(self, x):
+                )
+                compiled_out = loaded(*args)
+                self.assertEqual(eager_out, compiled_out)
+
```
@zou3519 do you want these tests to be done once or committed as unit tests?

zou3519 replied:
committed as unit tests. It does not matter to me if you do this before or after landing the diff but since they're tests for this diff putting them in the diff sounds easier?

zou3519 reviewed (approved):
LGTM, thank you

oulgen committed:
None on "Add inductor standalone_compile API"


This PR adds standalone_compile API that does precompilation via caching to support vLLM use case in the short term while we work on the longer term precompilation solution.

```
standalone_compile(gm, example_inputs, options) -> CompiledArtifact
CompiledArtifact.save(path, format: binary|unpacked = binary)
CompiledArtifact.load(path, format: binary|unpacked = binary)
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

oulgen committed:
None on "Add inductor standalone_compile API"


This PR adds standalone_compile API that does precompilation via caching to support vLLM use case in the short term while we work on the longer term precompilation solution.

```
standalone_compile(gm, example_inputs, options) -> CompiledArtifact
CompiledArtifact.save(path, format: binary|unpacked = binary)
CompiledArtifact.load(path, format: binary|unpacked = binary)
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

oulgen committed:
None on "Add inductor standalone_compile API"


This PR adds standalone_compile API that does precompilation via caching to support vLLM use case in the short term while we work on the longer term precompilation solution.

```
standalone_compile(gm, example_inputs, options) -> CompiledArtifact
CompiledArtifact.save(path, format: binary|unpacked = binary)
CompiledArtifact.load(path, format: binary|unpacked = binary)
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

zou3519 commented on .ci/docker/requirements-ci.txt:
```diff
@@ -379,3 +379,6 @@ dataclasses_json==0.6.7
 #test that import:
+
+# tests
+multiprocess
```
Why not use the built-in multiprocessing package?

oulgen replied:
multiprocessing uses pickle, and i was getting unable to pickle errors, not telling me what exactly. Swapping to multiprocessing which uses dill fixed it. This is only a ci dependency, so it should be ok, no?

zou3519 replied:
When I wrote the comment I was thinking about something more like [this](https://github.com/pytorch/pytorch/blob/main/test/test_pytree.py#L886), where we shell out a new script that loads from the directory, passes it a Tensor or some tensors (that are not created randomly), and then does an assertion on the output.

Adding a CI dependency is fine, it's just a bit annoying to deal with (you may need to update how packages are installed in CI, and local developers now need to know they need to install said package for testing)

youkaichao commented on torch/_inductor/standalone_compile.py:
```diff
@@ -0,0 +1,185 @@
+    def __call__(self, *args: Any) -> Any:
+        return self._compiled_fn(*args)[0]
+
+    def save(
```
`def save(self, path)` has two drawbacks:

1. the `path` is absolute path, and it cannot be reused across machines if the cache directory change.
2. as a user of the function, vLLM needs to generate a unique path for different shape, since we can call inductor multiple times.

we can `def save(self, format) -> str`, where the returned string is a key we use to `load` later. inductor can remember what does this key mean, inside its cache directory.

zou3519 replied:
@youkaichao I was thinking that vLLM can programmatically decide where it wants to save and load the new compiled artifact. 

For example, when compiling for batch size 0, assuming the hash is adsfasdf, vLLM can specify the path to be `{current_vllm_cache_dir}/torch_compile_cache/adsfasdf/batch_size0`.

When loading the compiled artifact for batch_size0, vLLM can specify the path to be`{current_vllm_cache_dir}/torch_compile_cache/adsfasdf/batch_size0`.

Do you think this isn't sufficient?

youkaichao replied:
okay, that should work. we will need wrap this logic in vLLM's compiler interface then. vLLM's compiler interfaces are `compile` and `load`, see https://github.com/vllm-project/vllm/blob/main/vllm/compilation/compiler_interface.py

oulgen committed:
None on "Add inductor standalone_compile API"


This PR adds standalone_compile API that does precompilation via caching to support vLLM use case in the short term while we work on the longer term precompilation solution.

```
standalone_compile(gm, example_inputs, options) -> CompiledArtifact
CompiledArtifact.save(path, format: binary|unpacked = binary)
CompiledArtifact.load(path, format: binary|unpacked = binary)
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

oulgen commented on test/inductor/test_codecache.py:
```diff
@@ -1327,6 +1330,131 @@ def forward(self, x):
+    @config.patch({"fx_graph_cache": True})
+    @config.patch({"fx_graph_remote_cache": False})
+    @functorch_config.patch({"enable_autograd_cache": True})
+    def test_different_process(self):
```
@zou3519 updated the test

zou3519 replied:
Looks good, thank you

oulgen committed:
None on "Add inductor standalone_compile API"


This PR adds standalone_compile API that does precompilation via caching to support vLLM use case in the short term while we work on the longer term precompilation solution.

```
standalone_compile(gm, example_inputs, options) -> CompiledArtifact
CompiledArtifact.save(path, format: binary|unpacked = binary)
CompiledArtifact.load(path, format: binary|unpacked = binary)
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

oulgen committed:
None on "Add inductor standalone_compile API"


This PR adds standalone_compile API that does precompilation via caching to support vLLM use case in the short term while we work on the longer term precompilation solution.

```
standalone_compile(gm, example_inputs, options) -> CompiledArtifact
CompiledArtifact.save(path, format: binary|unpacked = binary)
CompiledArtifact.load(path, format: binary|unpacked = binary)
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

oulgen commented:
@pytorchbot merge

albanD commented:
@pytorchbot revert -m "Broke profiler test" -c nosignal

The test has been consistently failing since this was merged see https://hud.pytorch.org/pytorch/pytorch/commit/bbc5fe850454df6860814ab77a1f3a4ca3698157

oulgen committed:
None on "Add inductor standalone_compile API"


This PR adds standalone_compile API that does precompilation via caching to support vLLM use case in the short term while we work on the longer term precompilation solution.

```
standalone_compile(gm, example_inputs, options) -> CompiledArtifact
CompiledArtifact.save(path, format: binary|unpacked = binary)
CompiledArtifact.load(path, format: binary|unpacked = binary)
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

oulgen commented:
@pytorchbot merge

Camyll commented:
@pytorchbot revert -m "breaking internal builds with torch module not found error" -c ghfirst

oulgen committed:
None on "Add inductor standalone_compile API"


This PR adds standalone_compile API that does precompilation via caching to support vLLM use case in the short term while we work on the longer term precompilation solution.

```
standalone_compile(gm, example_inputs, options) -> CompiledArtifact
CompiledArtifact.save(path, format: binary|unpacked = binary)
CompiledArtifact.load(path, format: binary|unpacked = binary)
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

oulgen commented:
@pytorchbot merge
